src/libmongoc/tests/test-mongoc-dns.c

```diff
@@ -728,6 +728,117 @@ prose_test_9_pooled (void *unused)
+   mongoc_apm_callbacks_t *cbs;
+   cb_stats_t stats = {0};
+
+   /* TXT record for test20.test.build.10gen.cc resolves to "loadBalanced=true". */
+   uri = mongoc_uri_new ("mongodb+srv://test20.test.build.10gen.cc/?replicaSet=rs0");
+   pool = mongoc_client_pool_new (uri);
+   cbs = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_topology_opening_cb (cbs, invalid_topology_opening);
+   mongoc_apm_set_topology_closed_cb (cbs, invalid_topology_closed);
+   mongoc_apm_set_server_opening_cb (cbs, invalid_server_opening);
+   mongoc_apm_set_server_closed_cb (cbs, invalid_server_closed);
+   mongoc_client_pool_set_apm_callbacks (pool, cbs, &stats);
+
+   ASSERT_CMPINT(stats.num_topology_opening, ==, 0);
+
+   /* Pop a client to attempt to start monitoring. Monitoring emits the topology_opening event on valid topologies. */
+   client = mongoc_client_pool_pop (pool);
+   mongoc_client_pool_push (pool, client);
+
+   mongoc_apm_callbacks_destroy (cbs);
+   mongoc_client_pool_destroy (pool);
+   mongoc_uri_destroy (uri);
+
+   ASSERT_CMPINT(stats.num_topology_opening, ==, 0);
+   ASSERT_CMPINT(stats.num_server_opening, ==, 0);
+   ASSERT_CMPINT(stats.num_server_closed, ==, 0);
+   ASSERT_CMPINT(stats.num_topology_closed, ==, 0);
+}
+
+/* CDRIVER-4184 Test that an invalid topology does not emit a topology_closed
+ * event. */
+static void
+test_invalid_topology_single (void *unused) {
+   mongoc_client_t *client;
+   mongoc_uri_t *uri;
+   mongoc_apm_callbacks_t *cbs;
+   cb_stats_t stats = {0};
+   bson_error_t error;
+   mongoc_server_description_t *sd;
+
+   /* TXT records for test20.test.build.10gen.cc resolve to loadBalanced=true. */
+   uri = mongoc_uri_new ("mongodb+srv://test20.test.build.10gen.cc/?replicaSet=true");
+   client = mongoc_client_new_from_uri (uri);
+   cbs = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_topology_opening_cb (cbs, invalid_topology_opening);
+   mongoc_apm_set_topology_closed_cb (cbs, invalid_topology_closed);
+   mongoc_apm_set_server_opening_cb (cbs, invalid_server_opening);
+   mongoc_apm_set_server_closed_cb (cbs, invalid_server_closed);
+   mongoc_client_set_apm_callbacks (client, cbs, &stats);
+
+   ASSERT_CMPINT(stats.num_topology_opening, ==, 0);
+
+   /* Perform server selection. Server selection emits the topology_opening
+    * event on valid topologies. */
+   sd = mongoc_client_select_server (
+      client, false /* for_writes */, NULL /* read_prefs */, &error);
+   ASSERT_ERROR_CONTAINS (
+      error,
+      MONGOC_ERROR_SERVER_SELECTION,
+      MONGOC_ERROR_SERVER_SELECTION_FAILURE,
+      "URI with \"loadbalanced\" enabled must not contain option \"replicaset\"");
+   ASSERT (!sd);
+
+   mongoc_apm_callbacks_destroy (cbs);
+   mongoc_client_destroy (client);
+   mongoc_uri_destroy (uri);
+
+   ASSERT_CMPINT(stats.num_topology_opening, ==, 0);
+   ASSERT_CMPINT(stats.num_server_opening, ==, 0);
+   ASSERT_CMPINT(stats.num_server_closed, ==, 0);
+   ASSERT_CMPINT(stats.num_topology_closed, ==, 0);
```

Comment:
...and this is the actual subject of the test? Is there a situation when it should be non-zero?

Reply:
Yes, this is the subject of the test. A `TopologyClosed` event is emitted in `mongoc_client_destroy` for a valid topology.

Reply:
Ok, I'm with you. This is is the only accepted result.